src/rgw/rgw_notify.cc

```diff
@@ -2,17 +2,530 @@
+    DECODE_START(1, bl);
+    decode(list, bl);
+    DECODE_FINISH(bl);
+  }
+};
+WRITE_CLASS_ENCODER(queues_t)
+
+class Manager {
+  const size_t max_queue_size;
+  const long queues_update_period_ms;
+  const long queues_update_retry_ms;
+  const long queue_idle_sleep_us;
+  const utime_t failover_time;
+  CephContext* const cct;
+  librados::IoCtx& rados_ioctx;
+  RGWUserPubSub ps_user;
+  std::unordered_set<std::string> owned_queues;
+  std::atomic_bool list_of_queues_object_created;
+  const std::string lock_cookie;
+  boost::asio::io_context io_context;
+  boost::asio::executor_work_guard<boost::asio::io_context::executor_type> work_guard;
+  const size_t worker_count;
+  std::vector<std::thread> workers;
+ 
+  const std::string Q_LIST_OBJECT_NAME = "queues_list_object";
+
+  // try create the object holding the list of queues if not created so far
+  // note that in case of race from two threads, or if another rgw
+  // already created the object, one would get -EEXIST, and the function would return 0 (success)
+  int try_to_create_queue_list() {
+    if (!list_of_queues_object_created) {
+      // create the object holding the list of queues
+      const auto ret = rados_ioctx.create(Q_LIST_OBJECT_NAME, false);
+      if (ret < 0 && ret != -EEXIST) {
+        ldout(cct, 1) << "ERROR: failed to create queue list. error: " << ret << dendl;
+        return ret;
+      }
+      list_of_queues_object_created = true;
+    }
+    return 0;
+  }
+
+  // read the list of queues from the queue list object
+  int read_queue_list(queues_t& queues) {
+    bufferlist bl;
+    constexpr auto chunk_size = 1024U;
+    auto start_offset = 0U;
+
+    int ret;
+    do {
+      bufferlist chunk_bl;
+      // TODO: add yield ?
+      ret = rados_ioctx.read(Q_LIST_OBJECT_NAME, chunk_bl, chunk_size, start_offset);
+      if (ret < 0) {
+        ldout(cct, 1) << "ERROR: failed to read queue list. error: " << ret << dendl;
+        return ret;
+      }
+      start_offset += ret;
+      bl.claim_append(chunk_bl);
+    } while (ret > 0);
+
+    if (bl.length() == 0) {
+      // nothing in the list
+      return 0;
+    }
+
+    auto iter = bl.cbegin();
+    try {
+      decode(queues, iter);
+    } catch (buffer::error& err) {
+      ldout(cct, 1) << "ERROR: failed to decode queue list. error: " << err.what() << dendl;
+      return -EINVAL;
+    }
+
+    return 0;
+  }
+
+  // set m1 to be the minimum between m1 and m2
+  static int set_min_marker(std::string& m1, const std::string m2) {
+    cls_queue_marker mr1;
+    cls_queue_marker mr2;
+    if (mr1.from_str(m1.c_str()) < 0 || mr2.from_str(m2.c_str()) < 0) {
+      return -EINVAL;
+    }
+    if (mr2.gen <= mr1.gen && mr2.offset < mr1.offset) {
+      m1 = m2;
+    }
+    return 0;
+  }
+
+   class tokens_waiter {
+     const boost::posix_time::time_duration infinite_duration = boost::posix_time::hours(1000);
+     size_t pending_tokens = 0;
+     boost::asio::deadline_timer timer;
```

Comment:
you can use `std::chrono` instead of `boost::posix_time` if you replace `deadline_timer` with `basic_waitable_timer<Clock>`. for example:
```c++
    using Clock = ceph::coarse_mono_clock;
    using Executor = boost::asio::io_context::executor_type;
    using Timer = boost::asio::basic_waitable_timer<Clock,
          boost::asio::wait_traits<Clock>, Executor>;
```